lte/gateway/python/magma/pipelined/app/enforcement.py

```diff
@@ -496,4 +496,192 @@ def deactivate_rules(self, imsi, ip_addr, rule_ids):
             self._deactivate_flows_for_subscriber(imsi, ip_addr)
         else:
             for rule_id in rule_ids:
-                self._deactivate_flow_for_rule(imsi, ip_addr, rule_id)
\ No newline at end of file
+                self._deactivate_flow_for_rule(imsi, ip_addr, rule_id)
+
+    def _install_default_flow_for_session(self, seid):
+        """
+        Add a low priority flow to drop traffic in the event
+        that all rules have been deactivated.
+
+        Args:
+            seid (string): session id
+        """
+        match = MagmaMatch(imsi=seid)
+        actions = []  # empty options == drop
+        flows.add_drop_flow(self._datapath, self.tbl_num, match, actions,
+                            priority=self.ENFORCE_DROP_PRIORITY)
+
+    def _install_flow_for_qos_rules_session(self, seid, pdrs, qers):
+        """
+        Install a flow to get stats for a particular rule. Flows will match on
+        seid, cookie (the rule num), in/out direction
+
+        Args:
+            seid (string): Session ID to install rule for
+            qers (QoS Enforcement Rules): QER for each PDR.
+                        pdrs : packet detection rules for uplink/downlink
+        """
+
+        flow_adds = []
+        try:
+            flow_adds = self._get_rule_match_flow_msgs_session(seid, qers, pdrs)
+        except FlowMatchError:
+            return RuleModResult.FAILURE
+
+        chan = self._msg_hub.send(flow_adds, self._datapath)
+        return self._wait_for_rule_responses_session(seid, chan)
+    def _get_rule_match_flow_msgs_session(self, seid, qers, pdrs):
+        """
+        Get flow msgs to get stats for a particular QER.
+        Each QER will match on seid, qer-id, cookie, in/out direction
+
+        Args:
+            seid (string): session ID to install rule for
+            qers (QoS Enforcement Rules): QER for each PDR.
+            pdrs : packet detection rules for uplink/downlink
+        """
+
+        flow_adds = []
+        for pdr in pdrs:
+            for qer in qers:
+                if (pdr.qer_id == qer.qer_id):
+                    priority = self.get_of_priority(pdr.precedence)
+                    try:
+                        flow_adds.extend(self._get_classify_rule_flow_msgs_session(seid, qer, priority))
+                    except FlowMatchError as err:
+                        self.logger.error(
+                                "Failed to get flow msg '%d' for session ID %s: %s",
+                                 qer.qer_id, seid, err)
+                        raise err
+                    break
+        return flow_adds
+
+    def _get_classify_rule_flow_msgs_session(self, seid, qer, priority):
+        """
+        Install a flow from a rule. If the flow action is close, then the flow
```

Comment:
This technically doen't install anything so lets clarify that it generates flow rule messages from the provided args

Reply:
I corrected the description of function.